### PR TITLE
[blog] Scoping in R (second attempt) + Four Kinds of Scoping in R

### PR DESCRIPTION
**Pull request by mhyee**
This is the second... and third... parts that supersedes #202. Originally this was just part two, but it got long enough that I split it off to make a "bonus" third part, that I plan to publish at the same time.

Most of the code examples are lifted straight from #202, but I added a few new examples, changed the metaprogramming examples, and significantly overhauled the discussion.

I still need to go back and add forward/backward links, update some of the text, update the dates, and fix up #203 (part one). But I think this draft is ready for feedback.

For convenience, I've rendered [part two](http://www.ccs.neu.edu/home/mhyee/scratch/Scoping%20in%20R.html?3) and [part three](http://www.ccs.neu.edu/home/mhyee/scratch/Four%20Kinds%20of%20Scoping%20in%20R.html?3).

I'd like to get feedback in by the end of this week, so I can post at the end of the week or early next week.

### CODE REVIEW

File: blog/_src/posts/2019-08-19-scoping-in-r.md [hunk 1]
```diff
@@ -0,0 +1,491 @@
...
+classify R as a lexically scoped language.
+
+Regardless, it is very easy to miscommunicate, so I think it's important to be
+very clear and make sure you and your audience know what definitions of scoping
+you're using!
+
+
+## Conclusion
+
+This entire adventure started when we were working on a paper[^4], and asked
```
Comment (bennn) [inline blog/_src/posts/2019-08-19-scoping-in-r.md]: `,[^4]`

File: blog/_src/posts/2019-08-19-scoping-in-r.md [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+and the creators have confirmed that lexical scoping was the intent.
+
+On the other hand, variable lookup depends on the run-time state of the
+program, and variable bindings cannot be resolved statically. Some of the
+examples even resemble dynamic scoping, where a free variable takes the value
+of the most recent assignment---but without scanning the call stack.
+Furthermore, the dynamic nature of R and its reflection and metaprogramming
+capabilities allow programmers to completely circumvent lexical scoping.
+
+This ambiguity shows up in a paper[^3], where the authors write:
```
Comment (bennn) [inline blog/_src/posts/2019-08-19-scoping-in-r.md]: move the footnote, to `,[^3]`

File: blog/_src/posts/2019-08-19-scoping-in-r.md [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+
+When we look up the value of `x` on line 5, R will first search the function's
+environment. If `b` evaluated to true, then R will find a value for `x`, which
+is `2`. Otherwise, R will search in the enclosing environment of `f`, and find
+that `x` is `1`.
+
+Both of these examples vaguely resemble dynamic scoping, in that `x` takes the
+value of the most recent assignment. However, this is not how R is
+implemented---as there is no traversal of the call stack---and it is not
+consistent with how R behaves in other examples.
```
Comment (bennn) [inline blog/_src/posts/2019-08-19-scoping-in-r.md]: The previous post doesn't say anything about traversing any call stack. Using those definitions, these examples are dynamic.

Maybe change to:
> Both of these example defy lexical scoping, in that ....
> However, the next section (link?) argues that R is lexically-scoped at heart.

File: blog/_src/posts/2019-08-19-scoping-in-r.md [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
+non-function values.
+
+In this example, we call `g` with the argument `42`, which is not a function.
+Then, in the body of `g`, we call `f(0)` on line 3, which requires looking up
+`f`. Although there is an `f` in the environment of `g`, its value is `42`,
+which is not a function. R will then search the enclosing environment, where it
+finds the function defined on line 1. Therefore, the lookup on line 3 resolves
+to the function on line 1, so `f(0)` returns `0`.
+
+This feature exists because `c` is the built-in function that constructs
```
Comment (bennn) [inline blog/_src/posts/2019-08-19-scoping-in-r.md]: `feature` -> `behavior`

File: blog/_src/posts/2019-08-19-scoping-in-r.md [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+  x <- 1
+  x <<- 2
+  x
+}
+f() # 1
+x   # 2
+```
+
+`<<-` is the "super assignment" operator. It skips the current environment and
+then searches the chain of enclosing environments until it finds a variable to
```
Comment (bennn) [inline blog/_src/posts/2019-08-19-scoping-in-r.md]: does it need to be a top level variable?

(I imagine this example would give the same result
```
x <- 0
f <- function() {
  x <- 1
  g <- function() {
    x <- 3
    x <<- 2
    x
  }
  g()
  x
}
f()
x 
```
)

File: blog/_src/posts/2019-08-19-scoping-in-r.md [hunk 6]
```diff
@@ -0,0 +1,491 @@
...
+When the body of `f` is evaluated, it needs to look up `x`. Lookup starts in
+`f`'s environment, which we set to be `enva`. Since no mapping for `x` is
+found, lookup continues in `enva`'s enclosing environment, which is `envb`.
+However, `envb` is also empty, so lookup continues in its enclosing
+environment, which is `enva`, and now lookup results in an infinite loop.
+
+
+### An intuition for scoping in R
+
+Some of the above examples appear to demonstrate dynamic scoping. Recall two of
```
Comment (bennn) [inline blog/_src/posts/2019-08-19-scoping-in-r.md]: appear to? hmmmm

File: blog/_src/posts/2019-08-19-scoping-in-r.md [hunk 7]
```diff
@@ -0,0 +1,491 @@
+    Title: Scoping in R
+    Date: 2019-08-19T15:15:57
+    Tags: scope, r, by Ming-Ho Yee
+
+In the previous post of this three-part blog series, we discussed lexical and
```
[the remark this answers is not in the document]
Comment (mhyee) [reply blog/_src/posts/2019-08-19-scoping-in-r.md]: Will do, once I know the URLs. (I'm going to update the dates so they match the publication date.)

File: blog/_src/posts/2019-08-19-scoping-in-r.md [hunk 8]
```diff
@@ -0,0 +1,491 @@
...
+dynamic scope. Now, in this second part, we can return to the original
+question: _is R lexically or dynamically scoped?_
+
+<!-- more -->
+
+Recall the "litmus test" from before:
```
[the remark this answers is not in the document]
Comment (mhyee) [reply blog/_src/posts/2019-08-19-scoping-in-r.md]: OK.

File: blog/_src/posts/2019-08-19-scoping-in-r.md [hunk 9]
```diff
@@ -0,0 +1,491 @@
...
+Let's examine what happens when we run this example. First, we create a mapping
+for `x` in the top-level environment. On line 2, we define a function `f`, which
+returns the value of some `x`. On line 3, we define a function `g`, which
+creates a new mapping for `x`, and then calls `f`. Note that the assignment on
+line 4 does _not_ update the definition on line 1.
+
+When `f` is called, it needs to look up the value of `x`. In other words, does
+the reference of `x` on line 2 refer to the assignment on line 1 or the
+assignment on line 4? If `f` returns `1`, then we know R is lexically scoped.
+If it returns `2`, then we know R is dynamically scoped.
```
[the remark this answers is not in the document]
Comment (mhyee) [reply blog/_src/posts/2019-08-19-scoping-in-r.md]: Good point, will reword.

File: blog/_src/posts/2019-08-19-scoping-in-r.md (hunk 1 above)
Comment (mhyee) [reply blog/_src/posts/2019-08-19-scoping-in-r.md]: Good catch, thanks.

File: blog/_src/posts/2019-08-19-scoping-in-r.md (hunk 3 above)
Comment (mhyee) [reply blog/_src/posts/2019-08-19-scoping-in-r.md]: Yeah, I'll have to think about this a bit. You're totally right that I never mentioned the call stack in the previous post, and I do want to keep the definition general. But I also don't want to say this example demonstrates dynamic scope. The lookup mechanism is definitely lexical-like, and if you take my intuition that it's _environments_ and not _variables_ that are being referenced, then this is lexically scoped.

File: blog/_src/posts/2019-08-19-scoping-in-r.md (hunk 4 above)
Comment (mhyee) [reply blog/_src/posts/2019-08-19-scoping-in-r.md]: I'll use `behaviour`. 😃

File: blog/_src/posts/2019-08-19-scoping-in-r.md (hunk 5 above)
Comment (mhyee) [reply blog/_src/posts/2019-08-19-scoping-in-r.md]: It'll skip the local variable and find the next one, so no, it doesn't have to be the top level. To annotate your example:

```
x <- 0
f <- function() {
  x <- 1
  g <- function() {
    x <- 3
    x <<- 2
    x # 3
  }
  g() # 3
  x # 2
}
f() # 2
x  # 0
```

And here's another example:
```
f <- function() {
  x <- 1
  x <<- 2
  x # 1
}
f() # 1
x # 2
```

File: blog/_src/posts/2019-08-19-scoping-in-r.md (hunk 6 above)
Comment (mhyee) [reply blog/_src/posts/2019-08-19-scoping-in-r.md]: Are you suggesting that I should strike out "appear to" and just say that the examples demonstrate dynamic scoping?

Again, I'm uncomfortable with that wording, even though it satisfies the definition I wrote in the previous post. That would mean R is both lexically and dynamically scoped, which is both confusing and wrong.

**Comment by mhyee** [comment]
Thanks @bennn! Did you get a chance to look over the third and bonus part? It's part of this PR. (It's OK if you're busy.)

> Did you read about [dynamic binding](https://en.wikipedia.org/wiki/Late_binding) while researching this post?
> 
> You may want to argue that R is function-scoped with dynamic binding.

So that's what I'm vaguely alluding to, but the problem is I didn't find a good definition of "dynamic binding" or "late binding." It's another vague term where everybody has a different definition (see discussion [here](https://ericlippert.com/2012/02/06/what-is-late-binding/)), and I don't like the Wikipedia definition. Wikipedia also uses it (along with [name binding](https://en.wikipedia.org/wiki/Name_binding)) mainly to describe dynamic dispatch.

(For that matter, I also dislike the Wikipedia article on scoping, because it mushes together different definitions, but at least I was able to find other sources.)

The only reasonable thing I found was in [The Art of the Interpreter](https://dspace.mit.edu/handle/1721.1/6094), pg 21, which says:

> We now construct an interpreter in which all variables have strictly local usage. This discipline is called _lexical scoping_ of variables, and has been used in many programming languages, including Algol 60. The term "lexical" refers to the fact that all references to a local variable binding are textually apparent in the program. The term _static binding_ is also used, indicating that the connection between binding and reference is unchanging at run time.

Which, unfortunately, would mean R is not lexically scoped. (Yes, this definition slightly contradicts the one Steele uses in _Common Lisp the Language_.)

> John Clements's dissertation might be helpful: https://www2.ccs.neu.edu/racket/pubs/dissertation-clements.pdf

I'll take a peek, thanks.

File: blog/_src/posts/2019-08-19-scoping-in-r.md (hunk 6 above)
Comment (bennn) [reply blog/_src/posts/2019-08-19-scoping-in-r.md]: It wasn't a suggestion, only a reaction.

**Comment by bennn** [comment]
I did look at third post, but didn't have anything to say.

File: blog/_src/posts/2019-08-19-scoping-in-r.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (mhyee) [reply blog/_src/posts/2019-08-19-scoping-in-r.md]:
```diff
 When `f` is called, it needs to look up the value of `x`. In other words, does
 the reference of `x` on line 2 refer to the assignment on line 1 or the
-assignment on line 4? If `f` returns `1`, then we know R is lexically scoped.
-If it returns `2`, then we know R is dynamically scoped.
+assignment on line 4? If `f` returns `1`, then the behaviour matches lexical
+scoping. If it returns `2`, then the behaviour matches dynamic scoping.
 
-When we run this example, the result is `1`. Therefore, R is lexically scoped.
+When we run this example, the result is `1`. This implies that R is lexically
+scoped.
```

File: blog/_src/posts/2019-08-19-scoping-in-r.md (hunk 3 above)
Comment (mhyee) [reply blog/_src/posts/2019-08-19-scoping-in-r.md]:
```diff
 Both of these examples vaguely resemble dynamic scoping, in that `x` takes the
-value of the most recent assignment. However, this is not how R is
-implemented---as there is no traversal of the call stack---and it is not
-consistent with how R behaves in other examples.
+value of the most recent assignment. However, this is not how R is implemented,
+and it is not consistent with how R behaves in other examples.
```

**Comment by mhyee** [comment]
I also revised the discussion section a bit:

```diff
 On the other hand, variable lookup depends on the run-time state of the
 program, and variable bindings cannot be resolved statically. Some of the
 examples even resemble dynamic scoping, where a free variable takes the value
-of the most recent assignment---but without scanning the call stack.
-Furthermore, the dynamic nature of R and its reflection and metaprogramming
-capabilities allow programmers to completely circumvent lexical scoping.
+of the most recent assignment---but this is not consistent with R's behaviour
+in other examples. Furthermore, the dynamic nature of R and its reflection and
+metaprogramming capabilities allow programmers to completely circumvent lexical
+scoping.
```

```diff
 Personally, I prefer the definition of _scope_ that treats name resolution as
 an orthogonal issue. I think it is more useful to keep the two issues separate.
+In addition, I think it is confusing and unhelpful to say that R is _both_
+lexically and dynamically scoped, or that R is _neither_ lexically and
+dynamically scoped.
```

Updated renders for [part two](http://www.ccs.neu.edu/home/mhyee/scratch/Scoping%20in%20R.html?3) and [part three](http://www.ccs.neu.edu/home/mhyee/scratch/Four%20Kinds%20of%20Scoping%20in%20R.html?3).

I've tentatively "scheduled" these for September 10. If I get more comments, I can update, but otherwise, I'll just push the merge button on the 10th.

**Pull request merged**